### ISSUE TITLE
TogglzRuleWithVariations_CalledMultipleTimes_Test.test is flaky

### ISSUE THREAD

**Issue by flakycov** [togglz/togglz#233]
`TogglzRuleWithVariations_CalledMultipleTimes_Test-test` failed for us on commit `bdb5c22` but did not fail on [the corresponding run on Travis](https://travis-ci.org/togglz/togglz/jobs/227665043).

Our tool confirms this test to be flaky. The stack trace is below. We set up our build environment to be as close as possible to your Travis build environment, using Ubuntu 12.04.5 LTS, Java 1.7.0_76 and Maven 3.3.9.

```
Tests run: 1, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 0.021 sec <<< FAILURE!
test(org.togglz.junit.vary.TogglzRuleWithVariations_CalledMultipleTimes_Test)  Time elapsed: 0.02 sec  <<< FAILURE!
java.lang.AssertionError: C3, F2 expected:<false> but was:<true>
	at org.junit.Assert.fail(Assert.java:93)
	at org.junit.Assert.failNotEquals(Assert.java:647)
	at org.junit.Assert.assertEquals(Assert.java:128)
	at org.togglz.junit.vary.TogglzRuleWithVariations_CalledMultipleTimes_Test.test(TogglzRuleWithVariations_CalledMultipleTimes_Test.java:39)
```

**Comment by chkal** [comment]
Thanks for reporting. I don't have any idea what is causing this. I guess we need to analyze this further.

**Comment by flakycov** [comment]
@chkal Thanks for your response! 

Is there some logging/debugging info that you think could help with debugging?

**Comment by chkal** [comment]
I guess we need to have a closer look at the test itself. Maybe we need to adjust the assertion to give us more information about the failure.

**Comment by flakycov** [comment]
Hi,
We found the cause for flakiness; while `getVariants` uses a LinkedHashSet to preserve the order of feature sets, `deepCopy` copies the set to a Hashset which doesn't preserve the order.  The fix is in the `deepCopy` method of `junit/src/main/java/org/togglz/junit/vary/VariationSetBuilder.java` - patch `copy = new Hashset...` to `copy = new LinkedHashSet...`

Let us know if you want us to open a PR.

We found this flaky test using DeFlaker (http://www.deflaker.org), and we used NonDex (https://github.com/TestingResearchIllinois/NonDex) to help in debugging.  DeFlaker also found an old flaky test (`SystemPropertyActivationStrategyTest.shouldBeFalseIfPropertyDoesNotExist`) that had been fixed meanwhile, and it found some more likely flaky tests that we don't have time to debug right now.  Let us know if you want to add any of these tools to your build.

Hope this helps.
